content/docs/iac/using-pulumi/extending-pulumi/build-a-provider.md

```diff
@@ -0,0 +1,756 @@
+	path := req.ID
+	byteContent, err := os.ReadFile(path)
+	if err != nil {
+		return infer.ReadResponse[FileArgs, FileState]{}, err
+	}
+	content := string(byteContent)
+	return infer.ReadResponse[FileArgs, FileState]{
+		ID: path,
+		Inputs: FileArgs{
+			Path:    path,
+			Force:   req.Inputs.Force && req.State.Force,
+			Content: content,
+		},
+		State: FileState{
+			Path:    path,
+			Force:   req.Inputs.Force && req.State.Force,
+			Content: content,
+		},
+	}, nil
+}
+
+func (*File) WireDependencies(f infer.FieldSelector, args *FileArgs, state *FileState) {
+	f.OutputField(&state.Content).DependsOn(f.InputField(&args.Content))
+	f.OutputField(&state.Force).DependsOn(f.InputField(&args.Force))
+	f.OutputField(&state.Path).DependsOn(f.InputField(&args.Path))
+}
+```
+
+We'll go through this code in detail in a bit, but for now, let's give it a try in a Pulumi program.
+
+{{% /choosable %}}
+
+{{< /chooser >}}
+
+### Use the provider in a Pulumi program
+
+Let's create a minimal Pulumi program that uses our new provider:
+
+```sh
+$ cd ..
+$ mkdir use-file-provider
+$ cd use-file-provider
+$ pulumi new yaml
+```
+
+This will initiatize a minimal YAML program.
+
+Edit the `Pulumi.yaml` file to have the following contents:
+
+```yaml
+name: use-file-provider
+runtime: yaml
+
+plugins:
+  providers:
+    - name: file
+      path: ../file-provider
+
+resources:
+  managedFile:
+    type: file:File
+    properties:
+      path: ${pulumi.cwd}/managed.txt
+      content: |
+        An important piece of information
+```
+
+Save that and then run `pulumi up`:
+
+```sh
+$ pulumi up
+...
+
+     Type                 Name                   Status
+ +   pulumi:pulumi:Stack  use-file-provider-dev  created (2s)
+ +   └─ file:index:File   managedFile            created (0.06s)
+
+Resources:
+    + 2 created
+```
+
+If all went well, you should see a new file created called `managed.txt` with the contents `An important piece of information`.
+
+You can verify this using `cat`:
+
+```sh
+$ cat managed.txt
+An important piece of information
+```
+
+### Detailed breakdown of provider implementation
+
+{{< chooser language "go" >}}
+
+{{% choosable language go %}}
+
+#### Preample and dependencies
+
+Like any other Go langauge module, you start with a `package` declaration and and `import` block. Here we are adding a few important packages from the base library (`context`, `fmt`, and `os`) which will help us with file operations, and a selection of imports from the Pulumi Provider SDK.
+
+```go
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	p "github.com/pulumi/pulumi-go-provider"
+	"github.com/pulumi/pulumi-go-provider/infer"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+	"github.com/pulumi/pulumi/sdk/v3/go/property"
+)
+```
+
+#### Provider entrypoint and indentity definitions
+
+The next section involves defining the `main()` entrypoint function and the `provider()` constructor.
```

Comment:
Here's the PR that introduced a builder pattern: https://github.com/pulumi/pulumi-go-provider/pull/317